queue.go

```diff
@@ -57,50 +55,47 @@ type Consumer interface {
 	Close() error
 }
 
-// Producer wraps the producer implementation details
+// Producer wraps the producer implementation details. Producer implementations
+// must support sync and async production.
 type Producer interface {
-	model.BatchProcessor
+	// Produce produces N records. If the Producer is synchronous, waits until
+	// all records are produced, otherwise, returns as soon as the records are
+	// stored in the producer buffer, or when the records are produced to the
+	// queue if sync producing is configured.
+	// If the context has been enriched with metadata, each entry will be added
+	// as a record's header.
+	// Produce takes ownership of Record and any modifications after Produce is
+	// called may cause an unhandled exception.
+	Produce(ctx context.Context, rs ...Record) error
 	// Healthy returns an error if the producer isn't healthy.
 	Healthy(ctx context.Context) error
 	// Close closes the producer.
 	Close() error
 }
 
-// Topic represents a destination topic where to produce a message/record.
-type Topic string
-
-// TopicRouter is used to determine the destination topic for an model.APMEvent.
-type TopicRouter func(event model.APMEvent) Topic
+// Record wraps a record's value with the topic where it's produced / consumed.
+type Record struct {
+	// Topics holds the topic where the record will be produced.
+	Topic Topic
+	// Value holds the record's content. It must not be mutated after Produce.
+	Value []byte
+}
```

Comment:
(Followup): Do you think we should add a field for timestamp to record the time at which the event was published to the queue?

Reply:
What's the purpose of the timestamp field and how would this timestamp field be used? Is the goal to use it in `Config.Processor`?

Reply:
> What's the purpose of the timestamp field and how would this timestamp field be used?

One use case can be to get the queued delay (the time for which a record was in a queued state before being consumed).

Reply:
Would you mind opening an issue for this enhancement? It seems like something we may want to implement, but I'm reluctant to add a field there until we have a legitimate reason to.

Reply:
👍